File: IOS/EasyHero/EasyHero/Data/Remote/HeroDto.swift
//
//  HeroDto.swift
//  EasyHero
//
//  Created by Alumno on 12/07/25.
//

struct HeroDto: Decodable, Identifiable {
    let id: String // ← es String en el JSON
    let name: String
    let image: HeroImage
    let stats: HeroStatsDTO
    let biography: HeroBiography

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case image
        case stats = "powerstats"
        case biography
    }

    func toDomain() -> Hero {
        Hero(
            id: Int(id) ?? 0,
            name: name,
            fullName: biography.fullName,
            image: image.url,
            stats: stats.toDomain()
        )
    }
}

struct HeroImage: Decodable {
    let url: String
}

struct HeroBiography: Decodable {
    let fullName: String

    enum CodingKeys: String, CodingKey {
        case fullName = "full-name"
    }
}

struct HeroStatsDTO: Decodable {
    let intelligence: String
    let strength: String
    let speed: String
    let durability: String
    let power: String
    let combat: String

    func toDomain() -> HeroStats {
        HeroStats(
            intelligence: intelligence,
            strength: strength,
            speed: speed,
            durability: durability,
            power: power,
            combat: combat
        )
    }
}

File: Android/EasyHero/app/src/main/java/com/example/easyhero/home/data/HeroDto.kt
package com.example.easyhero.home.data

import com.example.easyhero.home.Hero
import com.example.easyhero.home.Powerstats
import com.google.gson.annotations.SerializedName

data class HeroDto(
    val id : Int,
    @SerializedName("name") val name: String,
    @SerializedName("biography") val biography: BiographyDto,
    @SerializedName("image") val image: ImageDto,
    @SerializedName("powerstats") val powerstats: PowerstatsDto
) {
    fun toDomain(): Hero {
        return Hero(
            id = id,
            name = name,
            realName = biography.fullName,
            imageUrl = image.url,
            powerstats = Powerstats(
                intelligence = powerstats.intelligence,
                strength = powerstats.strength,
                speed = powerstats.speed,
                durability = powerstats.durability,
                power = powerstats.power,
                combat = powerstats.combat
            )
        )
    }
}

data class BiographyDto(
    @SerializedName("full-name") val fullName: String
)

data class ImageDto(
    @SerializedName("url") val url: String
)

data class PowerstatsDto(
    @SerializedName("intelligence") val intelligence: String,
    @SerializedName("strength") val strength: String,
    @SerializedName("speed") val speed: String,
    @SerializedName("durability") val durability: String,
    @SerializedName("power") val power: String,
    @SerializedName("combat") val combat: String
)

data class HeroesDto(
    @SerializedName("response") val response: String,
    @SerializedName("results") val heroes: List<HeroDto>? = null,
    @SerializedName("error") val error: String? = null
)